internal/apis/config/controller/types.go

```diff
@@ -0,0 +1,220 @@
+/*
+Copyright 2021 The cert-manager Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"time"
+
+	"k8s.io/component-base/logs"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	//"k8s.io/kubectl/pkg/cmd/logs"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+type ControllerConfiguration struct {
```

Comment:
Do we need an internal type?
Also, if we have an internal config type, could we make its fields non-optional (because they will contain the default values instead)?

Reply:
The rationale for this can be found in the [versioned component documentation](https://docs.google.com/document/d/1FdaEJUEh091qf5B98HM6_8MS764iXrxxigNIdwHYW9c/edit?usp=sharing). Quoting from the doc: 

> In order to maintain command-line compatibility, it must still be possible to parse flags for a period of time after they become available in your config API. When a field moves from your flags structure to your config structure, you should update the corresponding flag registration to target the internal config structure. It helps to keep all flag registrations close to each other. 

Based on that, it appears that it isn't completely necessary, but will help with slowly moving over to a config file and still supporting the command line options. At least that is how I read it.